### PR TITLE
[GLUTEN-7685][VL][RAS] Add new cost model to avoid costly r2c

### PR DESCRIPTION
**Pull request by zml1206**
## What changes were proposed in this pull request?

(Fixes: \#7685)
In production, we encountered a situation where the consumption of r2c large table was far greater than the performance improvement brought by native.
for example:
```
sql("set spark.gluten.sql.columnar.filescan=false")
spark.range(100000000).toDF("id").selectExpr("concat('id_', round(id/1000000)) as k", "id % 10 as v")
      .write.mode("overwrite").parquet("tmp/t1")
spark.read.parquet("tmp/t1").createOrReplaceTempView("t1")
sql("select  k,sum(v) as v from t1 group by k").collect()
```

The local test takes 18 seconds to enble gluten, and only 6 seconds to disablegluten.Therefore, I hope to fallback this through RAS.

The optimization points are as follows:
1. Increase bytesSize factor, cost = bytesSizeFactor * opCost
2. r2c cost can be configured separately, and the default is 100. If sizeBytes is less than the threshold, the cost of RowToColumnarLike is ignored.
3. Remove r2c fallback containing complex types
4. Vanilla op cost is configurable, the default is 20, gluten op cost is 1

## How was this patch tested?

UT and local test，the picture below shows the tasks that have significantly improved our environment.

image

### CODE REVIEW

**Comment by zml1206** [comment]
CI failed unrelated.

**Comment by zhztheplayer** [comment]
Run Gluten Clickhouse CI

File: gluten-substrait/src/main/scala/org/apache/gluten/planner/cost/RoughCostModel.scala [hunk 1]
```diff
@@ -16,50 +16,62 @@
...
+import org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec
 
 /** A rough cost model with some empirical heuristics. */
 class RoughCostModel extends LongCostModel {
 
+  private def getSizeFactor(plan: SparkPlan): Long = {
+    // Get the bytes size that the plan needs to consume
+    val sizeBytes = plan match {
+      case scan: DataSourceScanExec => getStatSizeBytes(scan)
+      case _: LeafExecNode => 0L
```
Comment (zhztheplayer) [inline gluten-substrait/src/main/scala/org/apache/gluten/planner/cost/RoughCostModel.scala:34]: Why this line is needed?

File: gluten-substrait/src/main/scala/org/apache/gluten/planner/cost/RoughCostModel.scala [hunk 2]
```diff
@@ -16,50 +16,62 @@
...
+import org.apache.spark.sql.execution.{ColumnarToRowExec, DataSourceScanExec, LeafExecNode, ProjectExec, RowToColumnarExec, SparkPlan}
+import org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec
 
 /** A rough cost model with some empirical heuristics. */
 class RoughCostModel extends LongCostModel {
 
+  private def getSizeFactor(plan: SparkPlan): Long = {
+    // Get the bytes size that the plan needs to consume
+    val sizeBytes = plan match {
+      case scan: DataSourceScanExec => getStatSizeBytes(scan)
```
Comment (zhztheplayer) [inline gluten-substrait/src/main/scala/org/apache/gluten/planner/cost/RoughCostModel.scala:33]: Do we need to consider DSV2?

**Review by zhztheplayer — COMMENTED** [review]
Is it possible for you to share some test results about this change? Thanks!

File: gluten-substrait/src/main/scala/org/apache/gluten/planner/cost/RoughCostModel.scala (hunk 1 above)
Comment (zml1206) [reply gluten-substrait/src/main/scala/org/apache/gluten/planner/cost/RoughCostModel.scala:34]: Some LeafExecNode bytesSize is particularly large and inaccurate, such as ExecutedCommandExec, which is ignored here. In addition, the AdaptiveSparkPlanExec match is lost here. I will fix it immediately.

**Comment by zml1206** [comment]
> Is it possible for you to share some test results about this change? Thanks!

I ran it in our grayscale adhoc query environment yesterday, and the overall performance increased by 10%. @zhztheplayer

**Comment by zhztheplayer** [comment]
> > Is it possible for you to share some test results about this change? Thanks!
> 
> I ran it in our grayscale adhoc query environment yesterday, and executorCpuTime overall reduced by 10%. @zhztheplayer

Sounds great.

BTW, based on my impression RoughCostModel has been used by a few of users. If you want to continue working on the new costers, would you consider creating a new one like `RoughCostModel2`, with alias `rough2`, probably? So we don't break the current usages of `rough` model by accident. Once it gets mature rough, we can move the logics to a more standardized cost model then.

**Review by zhztheplayer — APPROVED** [review]
👍

**Comment by zhztheplayer** [comment]
@zml1206 Can you add one line in PR description to demonstrate the way to enable this cost model? Thanks.

**Comment by zml1206** [comment]
> @zml1206 Can you add one line in PR description to demonstrate the way to enable this cost model? Thanks.

Added.

**Pull request merged**